src/ztex_phpass.c

```diff
@@ -0,0 +1,258 @@
+/*
+ * This software is Copyright (c) 2018 Denis Burykin
+ * [denis_burykin yahoo com], [denis-burykin2014 yandex ru]
+ * and it is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * This file is part of John the Ripper password cracker,
+ * Copyright (c) 1996-2001,2008,2010-2013,2015 by Solar Designer
+ *
+ * This file is part of John the Ripper password cracker,
+ * Copyright (c) 2012 magnum / JimF
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * There's ABSOLUTELY NO WARRANTY, express or implied.
+ *
+ */
+#include <stdio.h>
+#include <string.h>
+
+#include "arch.h"
+#include "common.h"
+#include "formats.h"
+#include "memory.h"
+#include "config.h"
+#include "options.h"
+#include "phpass_common.h"
+
+#include "ztex/device_bitstream.h"
+#include "ztex/device_format.h"
+#include "ztex/task.h"
+#include "ztex/pkt_comm/cmp_config.h"
+
+
+#define FORMAT_LABEL		"phpass-ztex"
+#define ALGORITHM_NAME		"phpass ($P$ or $H$)"
+
+#define FORMAT_NAME		""
+
+#define BENCHMARK_COMMENT	""
+
+// Partial hash in the database - differs from CPU/GPU implementations
+#undef	BINARY_SIZE
+#define	BINARY_SIZE		4
+
+// Actual salt size is 9
+#define	SALT_SIZE		8
+#define	SALT_ALIGN		1
+
+
+static struct device_bitstream bitstream = {
+	// bitstream ID (check is performed by querying operating bitstream)
+	0xd5c0,
+	"$JOHN/ztex/ztex115y_md5crypt.bit",
+	// parameters for high-speed packet communication (pkt_comm)
+	{ 2, 14336, 4094 },
+	// computing performance estimation (in candidates per interval)
+	// (keys * mask_num_cand)/crypt_all_interval per jtr_device.
+	1,		// keys/fpga for crypt_all(): set by init()
+	4096,	// keys/fpga for self-test
+	1536 * 1024,	// Would be 48 MB of USB traffic on 32-byte keys
+	512,		// Max. number of entries in onboard comparator.
+	384,		// Min. number of keys for effective device utilization
+	1, { 135 },	// Programmable clocks
+	"phpass",	// label for configuration file
+	"\x01", 1	// Initialization data
+};
+
+
+static int target_rounds;
+
+static void init(struct fmt_main *fmt_main)
+{
+//printf("verbosity (init):%d\n", options.verbosity);
+	//
+	// It gets TargetRounds from john.conf and adjust
+	// bitstream.candidates_per_crypt.
+
+	// Starting from TARGET_ROUNDS_1KPC, it sets keys_per_crypt
+	// equal to bitstream.min_keys
+	//
+	const int TARGET_ROUNDS_1KPC = 512*1024;
+
+	target_rounds = cfg_get_int("ZTEX:", bitstream.label,
+			"TargetRounds");
+	if (target_rounds <= 0)
+		target_rounds = 2048;
+
+	if (target_rounds < 512) {
+		fprintf(stderr, "Warning: invalid TargetRounds=%d in john.conf."
+			" Valid values are 1000-999999999\n", target_rounds);
+		target_rounds = 512;
```

Comment:
The message is inconsistent with the check here, and only powers of 2 may be used with phpass, so perhaps the message should say 512-999999999?

Reply:
Ageed.